### ISSUE TITLE
从yaml文件开始训练yolov12n，训练集cls_loss一直290左右，验证集的R P map都一直为0

### ISSUE THREAD

**Issue by Twsaiae** [sunsmarterjie/yolov12#76]
是因为我的类别比较多吗，我有20个类，还是数据集比较小啊，才280张图。
Ultralytics 8.3.63 🚀 Python-3.11.9 torch-2.5.1+cu124 CUDA:0 (NVIDIA RTX 2000 Ada Generation Laptop GPU, 8188MiB)
engine\trainer: task=detect, mode=train, model=yolov12n_mine.yaml, data=dataset.yaml, epochs=600, time=None, patience=100, batch=32, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=0, project=None, name=train9, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=True, opset=None, workspace=None, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.0, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.1, copy_paste_mode=flip, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=E:\codes\yolov12\runs\detect\train9

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      2368  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2, 1, 2]          
  2                  -1  1      6640  ultralytics.nn.modules.block.C3k2            [32, 64, 1, False, 0.25]      
  3                  -1  1      9344  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2, 1, 4]          
  4                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  5                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  6                  -1  2    174720  ultralytics.nn.modules.block.A2C2f           [128, 128, 2, True, 4]        
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  2    677120  ultralytics.nn.modules.block.A2C2f           [256, 256, 2, True, 1]        
  9                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 10             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 11                  -1  1     86912  ultralytics.nn.modules.block.A2C2f           [384, 128, 1, False, -1]      
 12                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 13             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 14                  -1  1     24000  ultralytics.nn.modules.block.A2C2f           [256, 64, 1, False, -1]       
 15                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 16            [-1, 11]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 17                  -1  1     74624  ultralytics.nn.modules.block.A2C2f           [192, 128, 1, False, -1]      
 18                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 19             [-1, 8]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 20                  -1  1    378880  ultralytics.nn.modules.block.C3k2            [384, 256, 1, True]           
 21        [14, 17, 20]  1    434572  ultralytics.nn.modules.head.Detect           [20, [64, 128, 256]]          
YOLOv12n_mine summary: 497 layers, 2,523,564 parameters, 2,523,548 gradients, 6.0 GFLOPs

Freezing layer 'model.21.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks...
AMP: checks passed ✅
train: Scanning E:\datasets\ocr_dataset_BlueFilmCamera_OCR_NG\yolo_ocr_dataset\datasets\imgs\train.cache... 223 images, 0 backgrounds, 0 corrupt: 100%|██████████| 223/223 [00:00<?, ?it/s]
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, num_output_channels=3, method='weighted_average'), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))
val: Scanning E:\datasets\ocr_dataset_BlueFilmCamera_OCR_NG\yolo_ocr_dataset\datasets\imgs\val.cache... 55 images, 0 backgrounds, 0 corrupt: 100%|██████████| 55/55 [00:00<?, ?it/s]
Plotting labels to E:\codes\yolov12\runs\detect\train9\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000417, momentum=0.9) with parameter groups 121 weight(decay=0.0), 128 weight(decay=0.0005), 127 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to E:\codes\yolov12\runs\detect\train9
Starting training for 600 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/600      5.13G     0.2023      290.7     0.1499        305        640: 100%|██████████| 7/7 [00:52<00:00,  7.56s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.99s/it]
                   all         55        330          0          0          0          0
  0%|          | 0/7 [00:00<?, ?it/s]
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      2/600      5.13G     0.1532        289     0.1101        231        640: 100%|██████████| 7/7 [00:08<00:00,  1.25s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.72s/it]
                   all         55        330          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      3/600      5.13G     0.1314      293.4     0.1029        280        640: 100%|██████████| 7/7 [00:08<00:00,  1.17s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:10<00:00, 10.51s/it]
                   all         55        330          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      4/600      5.13G     0.1282      288.9    0.08478        237        640: 100%|██████████| 7/7 [00:09<00:00,  1.29s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:11<00:00, 11.10s/it]
                   all         55        330          0          0          0          0

**Comment by Twsaiae** [comment]
> Could you test on a larger dataset first to check the code and environment?

我做了这些内容：1、我是用YOLO11n训练相同的数据集，以相同的参数，YOLO11n正常收敛。
2、我在YOLOv12n上训练COCO8.yaml，以相同的参数，YOLOv12n正常收敛。

给我整不会了。。。

**Comment by partcompany** [comment]
我也碰到了这个问题，你解决了么，我数据集比较大。

**Comment by sunsmarterjie** [comment]
Please first make sure the code and env are no problem, and then try to use residual connections for all scale models.

**Comment by Mshir0** [comment]
setting amp False,the yolov12 training gradient unstable

**Comment by sujiu7892** [comment]
问题解决了吗？我用yolov12也是出现0.004mAP的情况，不知道是不是我的设置出现了潜在的问题，同样的代码在yolov11上运行就是正常的，我使用的是windows系统Python-3.10 torch-2.5.1+cu124，但是我正常下载了相应版本的flash_attn，目前没找到那儿出了问题，求帮助

**Comment by heyang1743** [comment]
我也是，我一共3000张数据，验证后R P map都是接近于0。yolov11就正常数据。

**Comment by heyang1743** [comment]
我已经找到解决办法了。并不是数据集的问题，是环境配置的问题。不要使用“flash_attn-2.6.3+cu123torch2.2.2cxx11abiFALSE-cp311-cp311-win_amd64.whl”这个文件，而要使用“flash_attn-2.7.0.post2-cp311-cp311-win_amd64.whl”。这个文件在https://github.com/huihui-support/flash-attention-windows/tree/main/cp311。
I have found a solution. It's not a problem with the data set, but a problem with the environment configuration.Do not use the file "flash_attn-2.6.3+cu123torch2.2.2cxx11abiFALSE-cp311-cp311-win_amd64.whl". Instead, "flash_attn-2.7.0.post2-cp311-cp311-win_amd64.whl" should be used.The file is available at https://github.com/huihui-support/flash-attention-windows/tree/main/cp311

**Comment by sujiu7892** [comment]
你好，之前我使用的是：flash_attn-2.7.4.post1+cu124torch2.5.1cxx11abiFALSE-cp310-cp310-win_amd64.whl  ，后来改成：flash_attn-2.7.0.post2+cu124torch2.5.1cxx11abiFALSE-cp310-cp310-win_amd64.whl
接近0mAP的情况仍然存在
Ultralytics 8.3.63 🚀 Python-3.10.0 torch-2.5.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24576MiB)
YOLOv12n summary (fused): 376 layers, 2,508,539 parameters, 0 gradients, 5.8 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:04<00:00,  2.34it/s]
                   all        591       1113    0.00284      0.452    0.00457    0.00153
Speed: 0.3ms preprocess, 1.1ms inference, 0.0ms loss, 2.1ms postprocess per image
Results saved to runs\detect\train

Hello, previously I used: flash_attn-2.7.4.post1+cu124torch2.5.1cxx11abiFALSE-cp310-cp310-win_amd64.whl ， Later changed to: flash.attn-2.7.0.post2+cu124torch 2.5.1cxx11abiFALSE-cp310-cp310-win.amd64.whl
The situation of approaching 0mAP still exists
Ultralytics 8.3.63 🚀 Python-3.10.0 torch-2.5.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24576MiB)
YOLOv12n summary (fused): 376 layers, 2,508,539 parameters, 0 gradients, 5.8 GFLOPs
Class Images Instances Box(P R mAP50 mAP50-95): 100%|██████████| 10/10 [00:04<00:00, 2.34it/s]
all 591 1113 0.00284 0.452 0.00457 0.00153
Speed: 0.3ms preprocess, 1.1ms inference, 0.0ms loss, 2.1ms postprocess per image
Results saved to runs\detect\train

**Comment by BakaQvO** [comment]
我也一样，同样的数据集同样的配置文件，v8是可以的但是12就mAP只有0.0009
